CHANGES.md

```diff
@@ -1,6 +1,361 @@
+* fixed a bug in which compiler-generated code would warn about lack of '?'
+
+Bug Fixes for Build Issues
+--------------------------
+* fixed a bug in which the c2chapel virtual environment was incorrectly built
+
+Bug Fixes for GPU Computing
+---------------------------
+
+Bug Fixes for Libraries
+-----------------------
+* fixed a crash when calling `.seek()` on a remote `fileReader`/`fileWriter`
+
+Bug Fixes for Tools
+-------------------
+* fixed a path parsing issue in `printchplbuilds.py`
+
+Bug Fixes for the Runtime
+-------------------------
+* fixed useable core determination when some PUs are inaccessible
+* fixed environment variable parsing for some runtime variables
+
+
+Third-Party Software Changes
+----------------------------
+* updated the bundled copy of LLVM to version 17.0.6
+* updated the Python package versions used by 'chpldoc'  
+  (see $CHPL_HOME/third-party/chpl-venv/chpldoc-requirements*.txt for details)
+
+Developer-oriented changes: Process
+-----------------------------------
+
+Developer-oriented changes: Documentation
+-----------------------------------------
+* updated some out-of-date details in frontend/lib/parser/README
+* made clarifications to the driver mode technical note  
+  (see https://chapel-lang.org/docs/2.0/technotes/driver.html)
+* added a note about using an older valgrind with newer LLVM versions
+  (see https://chapel-lang.org/docs/2.0/developer/bestPractices/Valgrind.html#llvm-compatibility)
+* added a note about debugging LLVM passes
+* removed the Docker packaging README in favor of Chapel's Dockerhub page  
+  (see https://hub.docker.com/r/chapel/chapel/)
+
+Developer-oriented changes: Syntactic / Naming Changes
+------------------------------------------------------
+
+Developer-oriented changes: Module changes
+------------------------------------------
+
+Developer-oriented changes: Performance improvements
+----------------------------------------------------
+
+Developer-oriented changes: Makefile / Build-time changes
+---------------------------------------------------------
+* fixed an issue preventing developer builds with `clang` versions prior to 16
+
+Developer-oriented changes: Compiler Flags
+------------------------------------------
+* made `--compiler-driver` the default and added `--no-compiler-driver`  
+  (see https://chapel-lang.org/docs/2.0/technotes/driver.html)
+* added a new flag, `--llvm-print-passes`
+  (see https://chapel-lang.org/docs/2.0/technotes/llvm.html#inspecting-individual-llvm-passes)
+* added more options for the `--llvm-print-ir-stage` flag
+
+Developer-oriented changes: Compiler improvements / changes
+-----------------------------------------------------------
+* improved support for `with` clauses on `foreach` loops
+* fixed a number of bugs with the compiler driver mode
+
+Developer-oriented changes: 'dyno' Compiler improvements / changes
+------------------------------------------------------------------
```

Comment:
Here is my suggestion:

```
* made numerous improvements to the 'dyno' resolver for types and calls:
  - implemented type resolution of module-level split-initialized variables
  - implemented `is [const] copyable` and `is [const] assignable` primitives
  - implemented `pragma "last resort"` function resolution logic
  - added support for param-folding `select`s w.r.t. copy elision & split-init
  - added support for resolving calls to the `this` method of a field
  - improved default initialization of generics
  - improved support for `borrowed` and `unmanaged` classes
  - improved support for interaction between variable arguments and tuples
  - fixed resolution of `eltType` queries on the `c_ptr` type
  - fixed a bug in resolving user-defined initializers of owned classes
  - eliminated incorrect error messages involving type queries
  - made calling `.type` on a type emit an error
* improved the prototype support for library files:
  - added the ability to reuse LLVM IR stored in a library file
  - changed to using an ID-based munging strategy for library files
  - resolve all concrete functions when generating a library file
```

I included most things as subbullets under the `made numerous improvements to the 'dyno' resolver for types and calls` since that's what they are about. I ordered these subbullets by new feature / improved feature / bug fix.